File: gateway/src/contracts.ts
import { createHash } from 'node:crypto'

import { z } from 'zod'

export const contractVersion = 'v1' as const
export const queueName = 'cloudrag-jobs'
export const promptTemplateVersion = '2026-05-12'
export const defaultEmbeddingModel = 'deterministic-hash-v1'

export const sourceMetadataSchema = z.record(z.string(), z.string()).default({})

export const ingestRequestSchema = z.object({
  sourceId: z.string().min(1).max(128).optional(),
  title: z.string().min(1).max(256).optional(),
  content: z.string().min(1),
  contentType: z.string().min(1).default('text/plain'),
  metadata: sourceMetadataSchema,
})

export type IngestRequest = z.infer<typeof ingestRequestSchema>

export const queryRequestSchema = z.object({
  prompt: z.string().min(1),
  topK: z.number().int().min(1).max(20).default(5),
  minScore: z.number().min(0).max(1).default(0.15),
  useCache: z.boolean().default(true),
})

export type QueryRequest = z.infer<typeof queryRequestSchema>

export const sourceReferenceSchema = z.object({
  sourceId: z.string(),
  chunkId: z.string(),
  documentHash: z.string(),
  score: z.number(),
  content: z.string(),
  title: z.string().nullable(),
  metadata: sourceMetadataSchema,
})

export const queryResponseSchema = z.object({
  answer: z.string(),
  cache: z.object({
    hit: z.boolean(),
    key: z.string(),
    similarity: z.number().nullable(),
  }),
  sources: z.array(sourceReferenceSchema),
  retrieval: z.object({
    topK: z.number().int(),
    tookMs: z.number().int(),
  }),
  requestId: z.string(),
  traceId: z.string(),
})

export type QueryResponse = z.infer<typeof queryResponseSchema>

export const ingestAcceptedResponseSchema = z.object({
  status: z.literal('accepted'),
  taskId: z.string(),
  sourceId: z.string(),
  documentHash: z.string(),
  queue: z.string(),
  requestId: z.string(),
  traceId: z.string(),
})

export type IngestAcceptedResponse = z.infer<typeof ingestAcceptedResponseSchema>

export const errorResponseSchema = z.object({
  error: z.object({
    code: z.string(),
    message: z.string(),
    details: z.unknown().optional(),
  }),
  requestId: z.string(),
  traceId: z.string(),
})

export type ErrorResponse = z.infer<typeof errorResponseSchema>

export const ingestJobPayloadSchema = ingestRequestSchema.extend({
  storage: z
    .object({
      kind: z.enum(['inline', 's3']),
      objectKey: z.string().optional(),
      bucket: z.string().optional(),
    })
    .default({ kind: 'inline' }),
  submittedAt: z.string().datetime(),
})

export const queryJobPayloadSchema = queryRequestSchema.extend({
  cacheKey: z.string(),
  collectionName: z.string(),
  embeddingModel: z.string(),
  promptTemplateVersion: z.string(),
  submittedAt: z.string().datetime(),
})

export const jobEnvelopeSchema = z.discriminatedUnion('jobName', [
  z.object({
    version: z.literal(contractVersion),
    jobName: z.literal('ingest.v1'),
    requestId: z.string(),
    traceId: z.string(),
    payload: ingestJobPayloadSchema,
  }),
  z.object({
    version: z.literal(contractVersion),
    jobName: z.literal('query.v1'),
    requestId: z.string(),
    traceId: z.string(),
    payload: queryJobPayloadSchema,
  }),
])

export type IngestJobEnvelope = Extract<
  z.infer<typeof jobEnvelopeSchema>,
  { jobName: 'ingest.v1' }
>
export type QueryJobEnvelope = Extract<
  z.infer<typeof jobEnvelopeSchema>,
  { jobName: 'query.v1' }
>

export const workerQueryResultSchema = queryResponseSchema.extend({
  requestId: z.string(),
  traceId: z.string(),
})

export type WorkerQueryResult = z.infer<typeof workerQueryResultSchema>

export function computeDocumentHash(content: string): string {
  return createHash('sha256').update(content, 'utf8').digest('hex')
}

export function buildCacheKey(input: {
  prompt: string
  topK: number
  minScore: number
  collectionName: string
  embeddingModel?: string
  promptTemplateVersion?: string
}): string {
  const normalizedPrompt = input.prompt.trim().replace(/\s+/g, ' ')
  const material = [
    'semantic-cache',
    contractVersion,
    normalizedPrompt,
    String(input.topK),
    input.minScore.toFixed(4),
    input.collectionName,
    input.embeddingModel ?? defaultEmbeddingModel,
    input.promptTemplateVersion ?? promptTemplateVersion,
  ].join('|')

  return `semantic-cache:${contractVersion}:${createHash('sha256').update(material).digest('hex')}`
}

File: workers/src/workers/contracts.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

CONTRACT_VERSION = "v1"
DEFAULT_EMBEDDING_MODEL = "text-embedding-3-small"
PROMPT_TEMPLATE_VERSION = "2026-05-12"


class IngestRequest(BaseModel):
    source_id: str = Field(alias="sourceId")
    title: str | None = None
    content: str
    content_type: str = Field(alias="contentType", default="text/plain")
    metadata: dict[str, str] = Field(default_factory=dict)


class StorageReference(BaseModel):
    kind: Literal["inline", "s3"] = "inline"
    object_key: str | None = Field(alias="objectKey", default=None)
    bucket: str | None = None


class IngestJobPayload(IngestRequest):
    storage: StorageReference = Field(default_factory=StorageReference)
    submitted_at: datetime = Field(alias="submittedAt")


class QueryRequest(BaseModel):
    prompt: str
    top_k: int = Field(alias="topK", default=5)
    min_score: float = Field(alias="minScore", default=0.15)
    use_cache: bool = Field(alias="useCache", default=True)


class QueryJobPayload(QueryRequest):
    cache_key: str = Field(alias="cacheKey")
    collection_name: str = Field(alias="collectionName")
    embedding_model: str = Field(alias="embeddingModel")
    prompt_template_version: str = Field(alias="promptTemplateVersion")
    submitted_at: datetime = Field(alias="submittedAt")


class IngestJobEnvelope(BaseModel):
    version: Literal["v1"]
    job_name: Literal["ingest.v1"] = Field(alias="jobName")
    request_id: str = Field(alias="requestId")
    trace_id: str = Field(alias="traceId")
    payload: IngestJobPayload


class QueryJobEnvelope(BaseModel):
    version: Literal["v1"]
    job_name: Literal["query.v1"] = Field(alias="jobName")
    request_id: str = Field(alias="requestId")
    trace_id: str = Field(alias="traceId")
    payload: QueryJobPayload


class ChunkRecord(BaseModel):
    chunk_id: str
    chunk_hash: str
    source_id: str
    document_hash: str
    title: str | None = None
    content: str
    content_type: str
    created_at: datetime
    metadata: dict[str, str] = Field(default_factory=dict)
    dense_vector: list[float]
    sparse_vector: dict[int, float]


class IngestionResult(BaseModel):
    status: Literal["processed"]
    source_id: str
    document_hash: str
    chunk_count: int
    duplicate_chunks: int
    object_key: str
    request_id: str
    trace_id: str


class SourceReference(BaseModel):
    source_id: str = Field(alias="sourceId")
    chunk_id: str = Field(alias="chunkId")
    document_hash: str = Field(alias="documentHash")
    score: float
    content: str
    title: str | None = None
    metadata: dict[str, str] = Field(default_factory=dict)


class CacheMetadata(BaseModel):
    hit: bool
    key: str
    similarity: float | None = None


class RetrievalMetadata(BaseModel):
    top_k: int = Field(alias="topK")
    took_ms: int = Field(alias="tookMs")


class QueryResult(BaseModel):
    answer: str
    cache: CacheMetadata
    sources: list[SourceReference]
    retrieval: RetrievalMetadata
    request_id: str = Field(alias="requestId")
    trace_id: str = Field(alias="traceId")


class SemanticCacheEntry(BaseModel):
    cache_key: str
    prompt: str
    query_embedding: list[float]
    top_k: int
    min_score: float
    collection_name: str
    embedding_model: str
    prompt_template_version: str
    response: QueryResult
    created_at: datetime
